plugins/autoconfigure/plugin.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2018-present the CoreDHCP Authors. All rights reserved
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package autoconfigure
+
+// This plugin implements RFC2563:
+// 1. If the client has been allocated an IP address, do nothing
+// 2. If the client has not been allocated an IP address
+//    (yiaddr=0.0.0.0), then:
+//    2a. If the client has requested the "AutoConfigure" option,
+//        then add the selected value and terminate processing
+//    2b. Otherwise, terminate processing and send no reply
+//
+// This plugin should be used at the end of the plugin chain,
+// after any IP address allocation has taken place.
+//
+// The optional argument is the string "DoNotAutoConfigure" or
+// "AutoConfigure" (or "0" or "1" respectively).  The default
+// is to DoNotAutoConfigure.
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/coredhcp/coredhcp/handler"
+	"github.com/coredhcp/coredhcp/logger"
+	"github.com/coredhcp/coredhcp/plugins"
+	"github.com/insomniacslk/dhcp/dhcpv4"
+)
+
+var log = logger.GetLogger("plugins/autoconfigure")
+
+var autoconfigure byte
+
+var Plugin = plugins.Plugin{
+	Name:   "autoconfigure",
+	Setup4: setup4,
+}
+
+var argMap = map[string]byte{
+	"0":                  0,
+	"1":                  1,
+	"DoNotAutoConfigure": 0,
+	"AutoConfigure":      1,
+}
+
+func setup4(args ...string) (handler.Handler4, error) {
+	if len(args) > 0 {
+		var ok bool
+		autoconfigure, ok = argMap[args[0]]
+		if !ok {
+			return nil, fmt.Errorf("unexpected value '%v' for autoconfigure argument", args[0])
+		}
+	}
+	if len(args) > 1 {
+		return nil, errors.New("too many arguments")
+	}
+	return Handler4, nil
+}
```

Comment:
Ugh. As far as I can see, that conflicts with section 2.1, which says:

> When a server responds with the value "AutoConfigure", the client MAY generate a link-local IP address if appropriate.  However, if the server responds with "DoNotAutoConfigure", the client MUST NOT generate a link-local IP address, possibly leaving it with no IP address.

And also section 2.2 (DHCP client behavior):

> When a DHCP OFFER is received, it is handled as described in [DHCP], section 4.4.1, with one exception.  If the 'yiaddr' field is 0x00000000, the Auto-Configure option must be consulted.  If this option is set to "AutoConfigure", then the DHCPOFFER MUST be ignored, and the DHCP client MAY generate a link-local IP address.  However, if this option is set to "DoNotAutoConfigure", then the DHCPOFFER MUST be ignored, and the client MUST NOT generate a link-local IP address.

From this, I inferred that it was possible for the DHCP server to respond with "AutoConfigure" as a way to signal to the client "make yourself a link local address straight away, don't hang around for multiple tries and timeouts"

To me the last para of 2.3 doesn't make any sense, since under what other circumstance could the server send a DHCPOFFER containing the option "AutoConfigure"?

Do you want a module flag (or dummy value like -1) for strict compliance with last para of 2.3? It would say "if yiaddr is 0.0.0.0 then abort". The difference with not enabling the plugin at all is, if you didn't enable the plugin then I think the 0.0.0.0 would drop through to the end and be sent to the client as-is (is that correct?)

Aside: I note the language for the first para in 2.3 is updated in https://www.rfc-editor.org/rfc/rfc8925#section-3.3.1 but it doesn't make any difference here (a coredhcp user is going to process the IPv6-Only Preferred option first)

Reply:
Ok yeah, paragraph 2.3 doesn't make much sense then. I guess the important bit is to not send the option to a client that doesn't support it, but this code already does the right thing there.

This is good as-is, no need to implement the behavior from 2.3 at the moment; we can revisit if someone reports issues with specific clients